### ISSUE TITLE
UX procedure for password change may lock out user when using password manager

### ISSUE THREAD

**Issue by hinnerkweiler** [friendica/friendica#7719]
Not a real Bug but could lock out a user if using password manager and so called random strong passwords like it happened to ... well... I am asking for a friend... :-)

In resetting password dialog enter a "New Password" and have a typo in "Repeat new password" and use the password manager to fill the current password (random 60 char from  database)

### Expected behavior
 - Have a warning about not matching entries for new password 
 - Prevent user from proceeding until typo is fixed 

### Actual behavior
 - No warning
 - page reloads 
 - [the Password manager asks to update the new password in it's database - OK]
 - Popup informs that password change failed due to not matching passwords

The current situation resulted in loosing the first password. Without this password no second attempt can be made to change that password (old password is now unknown) and retrieving/resetting the old password fails due to https://github.com/friendica/friendica/issues/7709   

### Steps to reproduce the problem
see above

### Friendica version you encountered the problem
2019.04

**Comment by tobiasd** [comment]
So the password manager of your friend did not manage the fill in into the form and they made a typo before copying the random string into the repeat field, but the password manager cleared the copied string and there was no backup of the random string that was used before?

Sounds like your friend needs some JavaScript to watch the input of the two field before enabling the submit button of the form.

**Comment by hinnerkweiler** [comment]
No, the supposed new password was manually entered and the pasword manager just filled in the old password. - This kind of stuff seems to only happen to humans... 

However the Information that the password was not changed due to a mismatch AFTER the reload is also very likely to be missed by a user since it is just a small black box in the upper corner of the screen and quite far away from the main page content (on lager screens). 

Preventing the whole from submission by some small JavaScript seems beneficial to me for that reason also.

**Comment by MrPetovan** [comment]
Yes, I agree that the confirmation messages aren't the most legible and their display could be improved.

However, in this particular case Friendica also has a reset password feature that allows to send a new randomly generated password to the account's email address.

**Comment by dew-git** [comment]
@MrPetovan I'm sorry, but passwords shouldn't be generated and sent to a user's email, in plaintext I assume.

**Comment by MrPetovan** [comment]
You aren't sorry, you're right and we welcome all contributions as this is an free and open-source project.

**Comment by dew-git** [comment]
@MrPetovan Sorry(actually) that sounded harsher than I thought. I'll think about making PR, I need to do some more digging through the code. I see a lot of MD5ing and `mt_rand`s that might be an issue. Along with some places that need HMACs.

**Comment by MrPetovan** [comment]
We've been introducing libsodium bindings and using `random_bytes` as much as possible. MD5 isn't used for authentication, the only remaining wart is the password generation routine indeed.

**Comment by rolandog** [comment]
Posting here because of indications from #7792; I don't fully understand the issue of not being able to change one's password:

- Login with current username and generated password (copied in clipboard from email without surrounding white space)
- Go to `Settings > Account > Password change`
- Paste current password in `Current password` field
- Paste new password (generated from password manager)
- Save settings
- Get notifications:
  - `Login failed`
  - `Password unchanged`

Could this be a regression?

**Comment by hinnerkweiler** [comment]
uh, it's a while... from what I remember the main issue was pointed out by @tobiasd: "Sounds like your friend needs some JavaScript to watch the input of the two field before enabling the submit button of the form." Because after submitting the form Safari overwrote the old password in it's database usiing the string that was typed in to the "new Password field". Because it was different from the "retype new password field" the server refuses the change like you pointed out. However, the previous password – autogenerated and always auto filled in – is now unknown to the keywordmanager but still in place at the server.   

So when the notification `Password unchanged` arrives at the user, the passwords are finally out of sync and the user can not log in anymore until the whole "lost password" process is reinitiated.

I know, this is a probably rare issue but could be addressed quite simple as pointed out above.   - However, I think it's not related to #7792


>
>    [...] 
>
>     * Paste new password (generated from password manager)
>
>     * _Retype new password_
> 
>     * Save settings **(unvalidated)**
> 
>     * Get notifications:
>       
>       * `Login failed`
>       * `Password unchanged`
> 
> 
> Could this be a regression?

**Comment by rolandog** [comment]
Thanks for your answer. Maybe I didn't clarify that I'm using an external password manager that doesn't integrate with the browser, neither does it auto-fill in anything (for reference, it is [Master Password](https://masterpassword.app/)).

Also, my browser --- Firefox 82.0b4 (64-bit) --- was not auto-filling fields, since I had not saved the generated password that I intended to change.

Perhaps I should file this as a separate bug?

ETA: The Friendica instance in which I am registered is using version `2020.09-1`

**Comment by MrPetovan** [comment]
Hi @rolandog , please file a separate issue if it isn't related to password field auto-filling.

**Comment by jluttine** [comment]
I also vote that the frontend javascript verifies that the two password fields are identical before submitting the form to the backend.
